### DISCUSSION TITLE
Issue composing FindOptions object

### DISCUSSION THREAD

**Discussion by maphe** [mikro-orm/mikro-orm#2971] [Q&A]
**Describe the bug**
When trying to build an `options` object outside of the repository find call, by composing the `populate` array, the type check complains about:
```
TS2322: Type 'string' is not assignable to type 'never'.
```
I can work around the compiler error by casting `as unknown` but then the ORM seems to just ignore the populate param.

**To Reproduce**
```ts
    const options: FindOptions<User> = { // also tried FindOptions<User, string> but same result
      fields: ['id'],
    }
    
    options.populate = ['address']; // <--- TS2322: Type 'string' is not assignable to type 'never'.
```
whereas
```
const options: FindOptions<User> = { fields: ['id'], populate = ['address'] };
```
works just fine.

**Expected behavior**
Allow building options object in multiple phases before passing to the repository call.

**Versions**

| Dependency | Version |
| - | - |
| node | 16 |
| typescript | 4.5.5 |
| mikro-orm | 5.0.5 |
| your-driver | mysql |

**Comment by B4nan** [comment]
I will reply more in detail tomorrow, but the problem is in the second generic type argument, you use the default which means no populate hint. We need this type arg to allow dot notation, and we need to keep this default otherwise it would always consider all relations as populated without the explicit hint.

**Reply by maphe** [reply]
So, does that mean we basically can't do anything about it?

**Reply by B4nan** [reply]
Well, even `const options: FindOptions<User> = { fields: ['id'], populate = ['address'] };` does not compile on my end. You simply need to provide the second type argument with a valid value, default means no populate hint as its `never`. Even this way it won't be inferred, as TS does not support partial type inference, you either infer both args or none. On the other hand, `const options: FindOptions<User, any> = {}` will work just fine. Passing just `string` is not enough as it apparently breaks the inference with string literals.

Alternatively you could cast the populate value you are assigning to `never[]`, so `options.populate = ['address'] as never[]`. 

Both are rather hacks, but I don't think we can keep the strict type safety while allowing this kind of composition without casting. What we could have is some builder for the `FindOptions`, with a fluent interface, this way we can probably mutate the type based on what you add via `asserts` type guards. I will need to experiment with this a bit, never used the `asserts` keyword before, but recently saw one library doing so.

**Reply by n4an** [reply]
is like my problem that i have 3 month ago, when you say to use never

`async findAllUsers<P extends string = never>  {}`

**Reply by maphe** [reply]
Ok, cool, thanks for looking into it, I understand the complications. Curious to see what you can pull off 👍
